Improved function widget

XutaxKamay opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://cutter.re/docs/contributing/code/getting-started.html) to this repository
- [x] I made sure to follow the project's [coding style](https://cutter.re/docs/contributing/code/development-guidelines.html)
- [x] I've updated the [documentation](https://cutter.re/docs/user-docs.html) with the relevant information (if needed)


**Detailed description**

Hello.

This is improving mainly the function widget by adding icons on the name column which can be colorized. 
However, there's still some work to do in my opinion:

- Adding function export functionality.
- Some functions on windows (or even in general) aren't marked as imports / entry point. 
- About the entry point it doesn't seem also to point to the real entry point.. I suppose "main" and entry-point is different which is a bit confusing to me.
- We could also add marking for TLS callbacks entries.
- Maybe improving the icons as my skills at Inkscape are really limited... I would be glad if someone could improve them if needed.
- When a function is highlighted, we might do something to the icon as well but I've no idea yet what.

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

**Test plan (required)**

I've tested it, it seems to work properly for now and doesn't seem to break anything but let me know if you find anything. (I've tested both vertical and horizontal view)

![Screenshot_20210405_163023](https://user-images.githubusercontent.com/20893801/113585428-7fb18580-962c-11eb-874d-7e533c778a3b.png)

<!-- What steps should the reviewer take to test your pull request? Demonstrate that the code is solid. Example: The exact actions you made and their outcome. Add screenshots/videos if the pull request changes UI. This is your time to re-check that everything works and that you covered all the edge cases -->


<!-- **Code formatting**
Make sure you ran clang-format on your code before making the PR. Check our contribution guidelines here: https://cutter.re/docs/contributing/code/getting-started.html -->

**Closing issues**
Closes #2598
<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if such). -->

karliss reviewed (changes requested):
In my opinion gradient doesn't help with readability and distinguishing between Fx Fi and FE with unique part already being small fraction of the whole icon.

karliss commented:
> About the entry point it doesn't seem also to point to the real entry point.. I suppose "main" and entry-point is different which is a bit confusing to me.

Yes entry-point and main are two different things. Main is concept of programming language, entry point is concept of executable format. During start executable might execute a bunch of code before main function.

XutaxKamay commented on src/widgets/FunctionsWidget.cpp:
```diff
@@ -172,13 +176,48 @@ QVariant FunctionModel::data(const QModelIndex &index, int role) const
-            const static QIcon importIcon(":/img/icons/import_light.svg");
-            return importIcon;
+    case Qt::DecorationRole: {
+        static QIcon icon(":/img/icons/function.svg");
+        static QIcon icon_imp(":/img/icons/function_import.svg");
+        static QIcon icon_entry(":/img/icons/function_entry.svg");
```
Well it's what was doing previous code, so I just took it as reference.
And alright I can remove the gradient by just doing pure black/white depending on the theme, but that might remove the dynamic color thing.

karliss replied:
> Well it's what was doing previous code

Just because old code did something, doesn't always mean that it was good code and more code should do the same.

XutaxKamay replied:
Alright okay, but then why it got accepted in the first place? This is a bit misleading.
Also, I'm not sure what you meant by your sentence about static functions variables.. It's either static functions or static variables?
Anyway, how do you want the icons being created? Do you want it inside the FunctionModel class?

karliss replied:
> Alright okay, but then why it got accepted in the first place? This is a bit misleading.

Any software project that has been maintained for extended period of time will have parts of varying quality. 

> Anyway, how do you want the icons being created? Do you want it inside the FunctionsWidget class?

Assuming they are used as DecorationRole having them as member variables in FunctionModel is fine.  Having it in FunctionsWidget if you are going to use it in FunctionModel doesn't make much sense since it would only introduce extra code for forwarding it to model. If you decide to use custom rendering instead of models DecorationRole that's a slightly different situation.

> for now I've just returned as a QIcon directly (you said to do that)

By returning directly i meant returning QIcon without manually converting it to pixmap.  From the looking at code I don't think QIcon as whole is cached that much. Qt has pixmap cache, but but that's for individual pixmaps, not for icons which  can consist of multiple pixmaps and svg images. Didn't see any icon cache. QICon::cacheKey seemed to be only used by QWindowsMenu to skip work when assigning the same icon and debug output. So it's probably better not to repeatedly create QIcon from file path.

XutaxKamay replied:
`Assuming they are used as DecorationRole having them as member variables in FunctionModel is fine. Having it in FunctionsWidget if you are going to use it in FunctionModel doesn't make much sense since it would only introduce extra code for forwarding it to model. If you decide to use custom rendering instead of models DecorationRole that's a slightly different situation.`

Oh right I don't even remember why I renamed FunctionModel to FunctionsWidget I was right before I edited the post. I'll do that then.

`Assuming they are used as DecorationRole having them as member variables in FunctionModel is fine. Having it in FunctionsWidget if you are going to use it in FunctionModel doesn't make much sense since it would only introduce extra code for forwarding it to model. If you decide to use custom rendering instead of models DecorationRole that's a slightly different situation.`

I'll solve that in the next force-push.

XutaxKamay commented on src/widgets/FunctionsWidget.cpp:
```diff
@@ -172,13 +176,48 @@ QVariant FunctionModel::data(const QModelIndex &index, int role) const
+        // so we can fill the icon to a color we want.
+        // I didn't know how to get the pixel size properly so did it this way,
+        // Since it is depending on the font size anyway.
+        QFontMetrics metrics(currentIndex == function_index ? highlightFont : defaultFont);
+        int size = metrics.height();
+
+        auto paintIcon = [&size](QIcon *icon, const QColor &color) {
+            QPixmap pixmap = icon->pixmap(size, size);
+            QPainter paint(&pixmap);
+
+            paint.setCompositionMode(QPainter::CompositionMode_SourceIn);
+            paint.fillRect(pixmap.rect(), color);
+            paint.end();
+
+            return pixmap;
+        };
```
About the Qt::white thing I didn't notice it, that's true it should have been just using `this->property("color")` but I've just totally forgot about it. Thanks.
The problem is that you can't colorize dynamically a QIcon (but apparently is with a QIconEngine, so I'll look on that) so you have to take the pixmap which results into using non-builtin logic. But apparently you're not for colorizing the icons.. I did that dynamically so it can respect the theme's current name functions, but you're against that and prefer to use two icons, one for light and dark theme if I understood you correctly ?

karliss replied:
I am not against the concept of dynamic color, but as you observed there is no easy way of doing that without throwing away some of Qt builtin sizing logic for handling svg icons and potentially making int pixelated or blurry on some configs.  Moving away from dynamic coloring was only 2/4 suggested approaches. It is partially up to you deciding how much you want to dig into this and if the other methods are worth the effort. My subjective opinion is that it's extra work with little gains taking into account  that color is already applied to rest of the row and the same information is already expressed using shape of the icon.

XutaxKamay replied:
Alright, well I'll do dark & light icons then. Thanks.

XutaxKamay commented:
> > About the entry point it doesn't seem also to point to the real entry point.. I suppose "main" and entry-point is different which is a bit confusing to me.
> 
> Yes entry-point and main are two different things. Main is concept of programming language, entry point is concept of executable format. During start executable might execute a bunch of code before main function.

Yep I know about it but I find it still confusing, I think main could be totally ignored and entry marked as green. Or we could do both with some others colors.

XutaxKamay commented:
I've moved back by doing icons for dark and light themes, I've moved entry-point to main (created another icon for it), and added some others icons for the future for TLS callbacks, exports and entry-point(s). (There can be multiple entry-points sometimes when someone is manually injecting a PE/ELF with multiple threads, but not much the topic anyway)

Here is my subjective view of how the colors should be on those:

- Exports: purple/rose
- Entry-point(s): red
- TLS callbacks: orange/yellow

ITAYC0HEN commented:
Hey, sorry for making you wait. I hope to find the time to go over it in the coming days and read the discussion :)
Regarding colors, I only briefly read it, but I prefer not see "Red" used in the colors as it is a color that is usually used negatively to mark an issue

xarkes reviewed (approved):
Sorry for the delay, honestly I like how this PR looks and I think we could merge it.

ITAYC0HEN reviewed (approved):
Looks okay indeed.
@karliss you left "changes required" so if you approve this PR we can merge

XutaxKamay commented:
Hey,

sorry I haven't had much time to do much work on it, so I left it as it is, especially if it looks good to you all.
